File: rust/src/statement.rs
use std::collections::BTreeSet;

use crate::amount::parse_amount_minor;
use crate::time::{format_rfc3339_millis, parse_rfc3339_millis};
use crate::{ValueError, ValueResult, required};

pub const MAX_STATEMENT_PAGE_SIZE: usize = 500;

#[must_use]
pub fn canonical_statement_timestamp(millis: i64) -> String {
    format_rfc3339_millis(millis)
}

/// # Errors
/// Returns an error when the timestamp is not an RFC 3339 instant.
pub fn normalize_statement_timestamp(value: &str) -> ValueResult<String> {
    parse_rfc3339_millis(value, "Value statement timestamp").map(format_rfc3339_millis)
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct StatementPosting {
    pub posting_id: String,
    pub transaction_id: String,
    pub account_id: String,
    pub account_sequence: String,
    pub posting_sequence: u64,
    pub asset: String,
    pub amount_minor: String,
    pub balance_before_minor: String,
    pub balance_after_minor: String,
    pub occurred_at: String,
    pub recorded_at: String,
    pub source_namespace: String,
    pub source_type: String,
    pub source_id: String,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct StatementCursor {
    pub account_sequence: String,
    pub posting_id: String,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct StatementPage {
    pub account_id: String,
    pub asset: String,
    pub previous_account_sequence: String,
    pub opening_balance_minor: String,
    pub closing_account_sequence: String,
    pub closing_balance_minor: String,
    pub entries: Vec<StatementPosting>,
    pub has_more: bool,
    pub next_cursor: Option<StatementCursor>,
}
/// # Errors
/// Returns `ValueError` when input violates an authority invariant.
#[allow(clippy::too_many_lines)]
pub fn build_value_statement_page(
    account_id: &str,
    asset: &str,
    previous_account_sequence: &str,
    opening_balance_minor: &str,
    limit: usize,
    postings: &[StatementPosting],
) -> ValueResult<StatementPage> {
    required(account_id, "Value statement account is required")?;
    required(asset, "Value statement asset is required")?;
    if !(1..=MAX_STATEMENT_PAGE_SIZE).contains(&limit) {
        return Err(ValueError::new("Value statement page size is invalid"));
    }
    let previous_sequence = parse_amount_minor(previous_account_sequence)?;
    if previous_sequence < 0 {
        return Err(ValueError::new(
            "Value statement account sequence is invalid",
        ));
    }
    let mut ordered = postings.to_vec();
    ordered
        .sort_by_key(|posting| parse_amount_minor(&posting.account_sequence).unwrap_or(i64::MIN));
    let mut expected = i128::from(previous_sequence) + 1;
    let mut balance = parse_amount_minor(opening_balance_minor)?;
    let mut posting_ids = BTreeSet::new();
    let mut sequences = BTreeSet::new();
    for posting in &ordered {
        required(&posting.posting_id, "Value statement posting is required")?;
        required(
            &posting.transaction_id,
            "Value statement transaction is required",
        )?;
        required(
            &posting.source_namespace,
            "Value statement source namespace is required",
        )?;
        required(
            &posting.source_type,
            "Value statement source type is required",
        )?;
        required(&posting.source_id, "Value statement source is required")?;
        for (timestamp, field) in [
            (&posting.occurred_at, "Value statement occurrence time"),
            (&posting.recorded_at, "Value statement record time"),
        ] {
            let parsed = parse_rfc3339_millis(timestamp, field)
                .map_err(|_| ValueError::new(format!("{field} must be a canonical timestamp")))?;
            if format_rfc3339_millis(parsed) != *timestamp {
                return Err(ValueError::new(format!(
                    "{field} must be a canonical timestamp"
                )));
            }
        }
        if posting.account_id != account_id {
            return Err(ValueError::new(
                "Value statement contains a foreign account",
            ));
        }
        if posting.asset != asset {
            return Err(ValueError::new(
                "Value statement contains a cross-asset posting",
            ));
        }
        if posting.posting_sequence == 0 || posting.posting_sequence > 9_007_199_254_740_991 {
            return Err(ValueError::new(
                "Value statement posting sequence is invalid",
            ));
        }
        let account_sequence = parse_amount_minor(&posting.account_sequence)?;
        if account_sequence <= 0 {
            return Err(ValueError::new(
                "Value statement account sequence is invalid",
            ));
        }
        if !posting_ids.insert(posting.posting_id.clone())
            || !sequences.insert(posting.account_sequence.clone())
        {
            return Err(ValueError::new(
                "Value statement contains a duplicate posting identity or sequence",
            ));
        }
        if i128::from(account_sequence) != expected {
            return Err(ValueError::new(
                "Value statement account sequence is not gap-free",
            ));
        }
        let amount_minor = parse_amount_minor(&posting.amount_minor)?;
        if amount_minor == 0 {
            return Err(ValueError::new("Value statement posting must be non-zero"));
        }
        let balance_before_minor = parse_amount_minor(&posting.balance_before_minor)?;
        let balance_after_minor = parse_amount_minor(&posting.balance_after_minor)?;
        if balance_before_minor != balance {
            return Err(ValueError::new(
                "Value statement balance-before continuity mismatch",
            ));
        }
        if i128::from(balance_after_minor) != i128::from(balance) + i128::from(amount_minor) {
            return Err(ValueError::new(
                "Value statement balance-after continuity mismatch",
            ));
        }
        balance = balance_after_minor;
        expected += 1;
    }
    let entries: Vec<_> = ordered.into_iter().take(limit).collect();
    let has_more = postings.len() > entries.len();
    let closing_sequence = entries.last().map_or_else(
        || previous_sequence.to_string(),
        |posting| posting.account_sequence.clone(),
    );
    let closing_balance = entries.last().map_or_else(
        || {
            parse_amount_minor(opening_balance_minor)
                .unwrap_or_default()
                .to_string()
        },
        |posting| posting.balance_after_minor.clone(),
    );
    let next_cursor = has_more
        .then(|| {
            entries.last().map(|posting| StatementCursor {
                account_sequence: posting.account_sequence.clone(),
                posting_id: posting.posting_id.clone(),
            })
        })
        .flatten();
    Ok(StatementPage {
        account_id: account_id.to_owned(),
        asset: asset.to_owned(),
        previous_account_sequence: previous_sequence.to_string(),
        opening_balance_minor: parse_amount_minor(opening_balance_minor)?.to_string(),
        closing_account_sequence: closing_sequence,
        closing_balance_minor: closing_balance,
        entries,
        has_more,
        next_cursor,
    })
}

File: typescript/src/statement.ts
import { parseAmountMinor } from "./amount.js";

export const MAX_STATEMENT_PAGE_SIZE = 500;

export interface ValueStatementPostingFact {
  readonly postingId: string;
  readonly transactionId: string;
  readonly accountId: string;
  readonly accountSequence: string;
  readonly postingSequence: number;
  readonly asset: string;
  readonly amountMinor: string;
  readonly balanceBeforeMinor: string;
  readonly balanceAfterMinor: string;
  readonly occurredAt: string;
  readonly recordedAt: string;
  readonly sourceNamespace: string;
  readonly sourceType: string;
  readonly sourceId: string;
}

export interface ValueStatementCursor {
  readonly accountSequence: string;
  readonly postingId: string;
}

export interface ValueStatementPage {
  readonly accountId: string;
  readonly asset: string;
  readonly previousAccountSequence: string;
  readonly openingBalanceMinor: string;
  readonly closingAccountSequence: string;
  readonly closingBalanceMinor: string;
  readonly entries: readonly ValueStatementPostingFact[];
  readonly hasMore: boolean;
  readonly nextCursor?: ValueStatementCursor;
}

function requireIdentity(value: string, field: string): void {
  if (value.trim().length === 0) throw new Error(`${field} is required`);
}

function parseSequence(value: string, allowZero: boolean): bigint {
  const sequence = parseAmountMinor(value);
  if (allowZero ? sequence < 0n : sequence <= 0n) {
    throw new Error("Value statement account sequence is invalid");
  }
  return sequence;
}

function requireTimestamp(value: string, field: string): void {
  const parsed = new Date(value);
  if (!Number.isFinite(parsed.getTime()) || parsed.toISOString() !== value) {
    throw new Error(`${field} must be a canonical timestamp`);
  }
}

export function buildValueStatementPage(input: {
  readonly accountId: string;
  readonly asset: string;
  readonly previousAccountSequence: string;
  readonly openingBalanceMinor: string;
  readonly limit: number;
  readonly postings: readonly ValueStatementPostingFact[];
}): ValueStatementPage {
  requireIdentity(input.accountId, "Value statement account");
  requireIdentity(input.asset, "Value statement asset");
  if (
    !Number.isSafeInteger(input.limit) ||
    input.limit < 1 ||
    input.limit > MAX_STATEMENT_PAGE_SIZE
  ) {
    throw new Error("Value statement page size is invalid");
  }

  const previousSequence = parseSequence(input.previousAccountSequence, true);
  let expectedSequence = previousSequence + 1n;
  let runningBalance = parseAmountMinor(input.openingBalanceMinor);
  let pageClosingSequence = previousSequence;
  let pageClosingBalance = runningBalance;
  const postingIds = new Set<string>();
  const accountSequences = new Set<string>();
  const ordered = [...input.postings].sort((left, right) => {
    const leftSequence = parseSequence(left.accountSequence, false);
    const rightSequence = parseSequence(right.accountSequence, false);
    return leftSequence < rightSequence
      ? -1
      : leftSequence > rightSequence
        ? 1
        : 0;
  });

  for (const [index, posting] of ordered.entries()) {
    requireIdentity(posting.postingId, "Value statement posting");
    requireIdentity(posting.transactionId, "Value statement transaction");
    requireIdentity(
      posting.sourceNamespace,
      "Value statement source namespace",
    );
    requireIdentity(posting.sourceType, "Value statement source type");
    requireIdentity(posting.sourceId, "Value statement source");
    requireTimestamp(posting.occurredAt, "Value statement occurrence time");
    requireTimestamp(posting.recordedAt, "Value statement record time");
    if (posting.accountId !== input.accountId) {
      throw new Error("Value statement contains a foreign account");
    }
    if (posting.asset !== input.asset) {
      throw new Error("Value statement contains a cross-asset posting");
    }
    if (
      !Number.isSafeInteger(posting.postingSequence) ||
      posting.postingSequence < 1
    ) {
      throw new Error("Value statement posting sequence is invalid");
    }
    const accountSequence = parseSequence(posting.accountSequence, false);
    if (
      postingIds.has(posting.postingId) ||
      accountSequences.has(posting.accountSequence)
    ) {
      throw new Error(
        "Value statement contains a duplicate posting identity or sequence",
      );
    }
    postingIds.add(posting.postingId);
    accountSequences.add(posting.accountSequence);
    if (accountSequence !== expectedSequence) {
      throw new Error("Value statement account sequence is not gap-free");
    }

    const amount = parseAmountMinor(posting.amountMinor);
    if (amount === 0n)
      throw new Error("Value statement posting must be non-zero");
    const balanceBefore = parseAmountMinor(posting.balanceBeforeMinor);
    const balanceAfter = parseAmountMinor(posting.balanceAfterMinor);
    if (balanceBefore !== runningBalance) {
      throw new Error("Value statement balance-before continuity mismatch");
    }
    if (balanceAfter !== balanceBefore + amount) {
      throw new Error("Value statement balance-after continuity mismatch");
    }

    runningBalance = balanceAfter;
    expectedSequence += 1n;
    if (index < input.limit) {
      pageClosingSequence = accountSequence;
      pageClosingBalance = balanceAfter;
    }
  }

  const entries = Object.freeze(ordered.slice(0, input.limit));
  const hasMore = ordered.length > entries.length;
  const lastEntry = entries.at(-1);
  return Object.freeze({
    accountId: input.accountId,
    asset: input.asset,
    previousAccountSequence: previousSequence.toString(),
    openingBalanceMinor: parseAmountMinor(input.openingBalanceMinor).toString(),
    closingAccountSequence: pageClosingSequence.toString(),
    closingBalanceMinor: pageClosingBalance.toString(),
    entries,
    hasMore,
    ...(hasMore && lastEntry !== undefined
      ? {
          nextCursor: Object.freeze({
            accountSequence: lastEntry.accountSequence,
            postingId: lastEntry.postingId,
          }),
        }
      : {}),
  });
}
